checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,363 @@
+    if (anno == null || AnnotationUtils.areSame(anno, BOTTOM)) {
+      return BOTTOM;
+    } else if (!AnnotationUtils.areSameByName(
+        anno, "org.checkerframework.checker.mustcall.qual.MustCall")) {
+      return anno;
+    }
+    List<String> values =
+        AnnotationUtils.getElementValueArray(anno, mustCallValueElement, String.class);
+    if (!values.contains("close")) {
+      return anno;
+    }
+    return createMustCall(values.stream().filter(s -> !"close".equals(s)).toArray(String[]::new));
+  }
+
+  /**
+   * Returns true iff the given element is a resource variable.
+   *
+   * @param elt an element; may be null, in which case this method always returns false
+   * @return true iff the given element represents a resource variable
+   */
+  private boolean isResourceVariable(@Nullable Element elt) {
+    return elt != null && elt.getKind() == ElementKind.RESOURCE_VARIABLE;
+  }
+
+  /** Treat non-owning method parameters as @MustCallUnknown when the method is called. */
+  @Override
+  public void methodFromUsePreSubstitution(ExpressionTree tree, AnnotatedExecutableType type) {
+    ExecutableElement declaration;
+    if (tree instanceof MethodInvocationTree) {
+      declaration = TreeUtils.elementFromUse((MethodInvocationTree) tree);
+    } else if (tree instanceof MemberReferenceTree) {
+      declaration = (ExecutableElement) TreeUtils.elementFromTree(tree);
+    } else {
+      throw new BugInCF("unexpected type of method tree: " + tree.getKind());
+    }
+    changeNonOwningParameterTypesToTop(declaration, type);
+    super.methodFromUsePreSubstitution(tree, type);
+  }
+
+  @Override
+  protected void constructorFromUsePreSubstitution(
+      NewClassTree tree, AnnotatedExecutableType type) {
+    ExecutableElement declaration = TreeUtils.elementFromUse(tree);
+    changeNonOwningParameterTypesToTop(declaration, type);
+    super.constructorFromUsePreSubstitution(tree, type);
+  }
+
+  /**
+   * Changes the type of each parameter not annotated as @Owning to top. Also replaces the component
+   * type of the varargs array, if applicable.
+   *
+   * <p>Note that this method is not responsible for handling receivers, which can never be owning.
+   *
+   * @param declaration a method or constructor declaration
+   * @param type the method or constructor's type
+   */
+  private void changeNonOwningParameterTypesToTop(
+      ExecutableElement declaration, AnnotatedExecutableType type) {
+    for (int i = 0; i < type.getParameterTypes().size(); i++) {
+      Element paramDecl = declaration.getParameters().get(i);
+      if (checker.hasOption(MustCallChecker.NO_LIGHTWEIGHT_OWNERSHIP)
+          || getDeclAnnotation(paramDecl, Owning.class) == null) {
+        AnnotatedTypeMirror paramType = type.getParameterTypes().get(i);
+        if (!paramType.hasAnnotation(POLY)) {
+          paramType.replaceAnnotation(TOP);
+        }
+        if (declaration.isVarArgs() && i == type.getParameterTypes().size() - 1) {
```

Comment:
It is not necessary to test `declaration.isVarArgs()` on every iteration.  Move this `if` statement after the `for` loop, which can then be changed from an indexed for loop to a foreach loop.

Reply:
You're right that this check can be moved outside the loop. However, this cannot be changed to a foreach loop, because it needs to iterate over two same-size lists (that of the parameters and that of the parameter types).